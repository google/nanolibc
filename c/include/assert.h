//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_ASSERT_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_ASSERT_H_

#include <sys/_cdefs.h>

__BEGIN_DECLS

void _nnlc_assert(int val, const char *expr, const char *fname, int line,
                  const char *func_name);

__END_DECLS

#define assert(expr)                                        \
  ({                                                        \
    const int __v_assert_val = (int)(expr);                 \
    _nnlc_assert(__v_assert_val, #expr, __FILE__, __LINE__, \
                 __PRETTY_FUNCTION__);                      \
  })

#if __STDC_VERSION__ >= 201112L
# define static_assert _Static_assert
#endif

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_ASSERT_H_
