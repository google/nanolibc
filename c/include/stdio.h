//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#ifndef THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_
#define THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_

#include <stdarg.h>
#include <sys/types.h>
#include <sys/_cdefs.h>

/* Some assume stdio.h defines NULL, etc. And SUSv2 is actually ok with
 * that: "Inclusion of the <stdio.h> header may also make visible all
 * symbols from <stddef.h>." */
#include <stddef.h>

__BEGIN_DECLS

#define EOF (-1)

typedef struct _FILE_DESCR FILE;

extern FILE *stdout;
extern FILE *stderr;

int vsnprintf(char *str, size_t size, const char *format, va_list ap);
int snprintf(char *str, size_t size, const char *format, ...)
    __attribute__((format(printf, 3, 4)));

int vsprintf(char *str, const char *format, va_list ap);
int sprintf(char *str, const char *format, ...)
    __attribute__((format(printf, 2, 3)));

int vfprintf(FILE *stream, const char *format, va_list ap);
int fprintf(FILE *stream, const char *format, ...)
    __attribute__((format(printf, 2, 3)));

int vprintf(const char *format, va_list ap);
int printf(const char *format, ...) __attribute__((format(printf, 1, 2)));

int fputc(int c, FILE *stream);
int fputs(const char *s, FILE *stream);
int putc(int c, FILE *stream);
int putchar(int c);
int puts(const char *s);

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream);

int fflush(FILE *stream);

int perror(const char *s);

/*
 * Published but functions NOT implemented!
 */
extern FILE *stdin;

#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

int scanf(const char *format, ...);
int sscanf(const char *str, const char *format, ...);
FILE *fopen(const char *path, const char *mode);
int fclose(FILE *fp);
char *fgets(char *s, int size, FILE *stream);
size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);
int fseek(FILE *stream, long offset, int whence);
long ftell(FILE *stream);
void rewind(FILE *stream);

__END_DECLS

#endif  // THIRD_PARTY_NANOLIBC_C_INCLUDE_STDIO_H_
