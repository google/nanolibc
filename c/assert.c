//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
 * Implementation of assert.h functions.
 */

#include <stdlib.h>
#include <stdio.h>

#include <assert.h>

/* assertion function called when #include <assert.h> was used */
void _nnlc_assert(int val, const char *expr, const char *fname, int line,
                  const char *funcname) {
  if (val) return;

  fprintf(stderr, "Assertion '%s' failed in %s (%s:%d)\n", expr, funcname,
          fname, line);
  abort();
}

/* for who doesn't #include <assert.h>, provide user with a default
 * assert() function to link with */
void _nanolibc_default_assert(int val) {
  if (val) return;

  fputs("Assertion failed!\n", stderr);
  abort();
}

#undef assert
void assert() __attribute__((weak, alias("_nanolibc_default_assert")));
