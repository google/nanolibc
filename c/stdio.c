//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*
 * Implementation of stdio.h functions.
 */

#include <stddef.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>
#include <stdio.h>

#include "third_party/nanolibc/c/libc_internals.h"
#include "third_party/tinyprintf/tinyprintf.h"

FILE *stdin = &__nnlc_internal_data.libc_stdin;
FILE *stdout = &__nnlc_internal_data.libc_stdout;
FILE *stderr = &__nnlc_internal_data.libc_stderr;

/* returns TRUE if param 'stream' is a valid FILE* pointer */
inline static int _is_valid_FILE(FILE *stream) {
  return ((stream != NULL) && (stream->magic == _NNLC_STDIO_MAGIC));
}

int fputc(int c, FILE *stream) {
  const unsigned char uc = c; /* according to man */
  if (!_is_valid_FILE(stream) || !stream->write) return EOF;
  if (1 != stream->write(&uc, 1)) return EOF;
  return uc; /* according to man */
}

int fputs(const char *s, FILE *stream) {
  size_t len;
  if (!_is_valid_FILE(stream) || !stream->write) return EOF;
  len = strlen(s);
  if (stream->write(s, len) != (ssize_t)len) return EOF;
  return 1;
}

int putc(int c, FILE *stream) { return fputc(c, stream); }

int putchar(int c) {
  const unsigned char uc = c; /* according to man */
  if (1 != stdout->write(&uc, 1)) return EOF;
  return uc; /* according to man */
}

int puts(const char *s) {
  const size_t len = strlen(s);
  if (stdout->write(s, len) != (ssize_t)len) return EOF;
  if (1 != stdout->write("\n", 1)) return EOF;
  return 1;
}

size_t fwrite(const void *ptr, size_t size, size_t nmemb, FILE *stream) {
  size_t i;
  if (!_is_valid_FILE(stream) || !stream->write) return 0;
  for (i = 0; i < nmemb; ++i) {
    const void *p = ptr + (i * size);
    if (stream->write(p, size) != (ssize_t)size) break;
  }
  return i;
}

int fflush(FILE *stream) {
  if (!_is_valid_FILE(stream)) return -1;

  return 0;
}

int perror(const char *s) {
  fprintf(stderr, "Error: %s\n", s);
  return 0;
}

/*
 * vprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to call the underlying runtime 'write'
 * functions. We use a small buffer on the stack to avoid calling the
 * underlying 'write' too often.
 */
struct _vprintf_putcf_data {
  FILE *stream;

  char *buff; /* allocated on the stack by vprintf */
  size_t buff_capacity;
  size_t nbuffered; /* slightly redundant with buff_capacity, we don't care */

  size_t total_written;
};

/* Callback function for tfp_format to add the given char c to the buffer */
static void _vprintf_putcf(void *p, char c) {
  struct _vprintf_putcf_data *data = (struct _vprintf_putcf_data *)p;

  /* sanity check */
  if (data->nbuffered >= data->buff_capacity)
    abort(); /* no assert(), to prevent infinite recursion */

  data->buff[data->nbuffered++] = c;

  /* need to flush buffer? */
  if (data->nbuffered >= data->buff_capacity) {
    int n = data->stream->write(data->buff, data->nbuffered);
    if (n > 0) data->total_written += n;
    data->nbuffered = 0;
  }
}

int vfprintf(FILE *stream, const char *format, va_list ap) {
  char buff[64]; /* any size >= 1 is Ok */
  struct _vprintf_putcf_data data;

  if (!_is_valid_FILE(stream) || !stream->write) return -1;

  data.stream = stream;
  data.buff_capacity = sizeof(buff);
  data.buff = buff;
  data.nbuffered = 0;
  data.total_written = 0;
  tfp_format(&data, _vprintf_putcf, format, ap);

  /* chars remaining in buffer? */
  if (data.nbuffered > 0) {
    int n = stream->write(data.buff, data.nbuffered);
    if (n > 0) data.total_written += n;
  }

  return data.total_written;
}

int fprintf(FILE *stream, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vfprintf(stream, format, ap);
  va_end(ap);
  return retval;
}

int vprintf(const char *format, va_list ap) {
  return vfprintf(stdout, format, ap);
}

int printf(const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vprintf(format, ap);
  va_end(ap);

  return retval;
}

/*
 * vsnprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to fill the buffer provided by caller.
 */
struct _vsnprintf_putcf_data {
  size_t dest_capacity;
  char *dest;
  size_t dest_used;
};

static void _vsnprintf_putcf(void *p, char c) {
  struct _vsnprintf_putcf_data *data = (struct _vsnprintf_putcf_data *)p;
  if (data->dest_used < data->dest_capacity) data->dest[data->dest_used] = c;

  data->dest_used++;
}

int vsnprintf(char *str, size_t size, const char *format, va_list ap) {
  struct _vsnprintf_putcf_data data;

  data.dest = str;
  data.dest_capacity = size;
  data.dest_used = 0;
  tfp_format(&data, _vsnprintf_putcf, format, ap);

  if (data.dest_used < size)
    data.dest[data.dest_used] = '\0';
  else if (size > 0)
    data.dest[size - 1] = '\0';

  return data.dest_used;
}

int snprintf(char *str, size_t size, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vsnprintf(str, size, format, ap);
  va_end(ap);
  return retval;
}

/*
 * vsprintf family of functions. Rely on tfp_format, provide required
 * callbacks and data for it to fill the buffer provided by caller.
 */
struct _vsprintf_putcf_data {
  char *dest;
  size_t dest_used;
};

static void _vsprintf_putcf(void *p, char c) {
  struct _vsprintf_putcf_data *data = (struct _vsprintf_putcf_data *)p;
  data->dest[data->dest_used++] = c;
}

int vsprintf(char *str, const char *format, va_list ap) {
  struct _vsprintf_putcf_data data;
  data.dest = str;
  data.dest_used = 0;
  tfp_format(&data, _vsprintf_putcf, format, ap);
  data.dest[data.dest_used] = '\0';
  return data.dest_used;
}

int sprintf(char *str, const char *format, ...) {
  va_list ap;
  int retval;

  va_start(ap, format);
  retval = vsprintf(str, format, ap);
  va_end(ap);
  return retval;
}
