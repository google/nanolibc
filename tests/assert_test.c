//  Copyright 2022 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/* Test assert() */

#include <stdio.h>
#include <assert.h>

static void test_macro_assert() {
  assert(1 != 0);
  assert(1 == 1);

  assert(2 == 3); /* expected to fail */

  fprintf(stderr, "You should never see this message!\n");
}

int main() {
  test_macro_assert();
  return 0;
}
