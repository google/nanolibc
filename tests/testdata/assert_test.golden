Assertion '2 == 3' failed in void test_macro_assert() (third_party/nanolibc/tests/assert_test.c:24)
# END TEST WITH RETVAL=255
